rustls-wolfcrypt-provider/src/verify/rsapss.rs

```diff
@@ -152,39 +176,86 @@ impl SignatureVerificationAlgorithm for RsaPssSha384Verify {
     }
 }
 
-fn wc_decode_spki_spk(spki_spk: &[u8]) -> Result<RsaKey, InvalidSignature> {
-    let mut reader = der::SliceReader::new(spki_spk).map_err(|_| InvalidSignature)?;
-    let ne: [der::asn1::UintRef; 2] = reader.decode().map_err(|_| InvalidSignature)?;
-    let n = BigUint::from_bytes_be(ne[0].as_bytes());
-    let e = BigUint::from_bytes_be(ne[1].as_bytes());
-    let n_bytes = n.to_bytes_be();
-    let e_bytes = e.to_bytes_be();
-
-    let mut rsa_key_c_type: RsaKey = unsafe { mem::zeroed() };
-    let rsa_key_object = unsafe { RsaKeyObject::from_ptr(&mut rsa_key_c_type) };
-    let mut ret;
-
-    // This function initializes a provided RsaKey struct. It also takes in a heap identifier,
-    // for use with user defined memory overrides (see XMALLOC, XFREE, XREALLOC).
-    ret = unsafe { wc_InitRsaKey(rsa_key_object.as_ptr(), ptr::null_mut()) };
-    check_if_zero(ret).unwrap();
-
-    // This function decodes the raw elements of an RSA public key, taking in
-    // the public modulus (n) and exponent (e). It stores these raw elements in the provided
-    // RsaKey structure, allowing one to use them in the encryption/decryption process.
-    ret = unsafe {
-        wc_RsaPublicKeyDecodeRaw(
-            n_bytes.as_ptr(),
-            n_bytes.capacity().try_into().unwrap(),
-            e_bytes.as_ptr(),
-            e_bytes.capacity().try_into().unwrap(),
-            rsa_key_object.as_ptr(),
-        )
-    };
-
-    if let Err(WCError::Failure) = check_if_zero(ret) {
-        Err(InvalidSignature)
-    } else {
-        Ok(rsa_key_c_type)
+#[derive(Debug)]
+pub struct RsaPssSha512Verify;
+
+impl SignatureVerificationAlgorithm for RsaPssSha512Verify {
+    fn public_key_alg_id(&self) -> AlgorithmIdentifier {
+        alg_id::RSA_ENCRYPTION
+    }
+
+    fn signature_alg_id(&self) -> AlgorithmIdentifier {
+        alg_id::RSA_PSS_SHA512
+    }
+
+    fn verify_signature(
+        &self,
+        public_key: &[u8],
+        message: &[u8],
+        signature: &[u8],
+    ) -> Result<(), InvalidSignature> {
+        let mut digest: [u8; WC_SHA512_DIGEST_SIZE as usize] = [0; WC_SHA512_DIGEST_SIZE as usize];
+        let mut out: [u8; 256] = [0; 256];
```

Comment:
Will this fail if used with 4096 bit RSA keys? Seems hard set to only handle 2048 bit signatures. If so can we make an adjustment here to handle the max RSA key size supported.